### PR TITLE
Improved Stream Router Engine

### PR DESCRIPTION
**Pull request by dennisoelkers**
This pull request adds the new stream router engine optimized for best guess minimum cost evaluation.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 1]
```diff
@@ -49,23 +49,13 @@
...
 
     private final List<Stream> streams;
-    private final StreamFaultManager streamFaultManager;
-    private final StreamMetrics streamMetrics;
-    private final TimeLimiter timeLimiter;
-    private final long streamProcessingTimeout;
+    protected final StreamFaultManager streamFaultManager;
+    protected final StreamMetrics streamMetrics;
+    protected final TimeLimiter timeLimiter;
+    protected final long streamProcessingTimeout;
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Why did you change the visibility of those final fields? They're only being accessed in this very class.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 2]
```diff
@@ -83,33 +73,49 @@ public StreamRouterEngine(@Assisted List<Stream> streams,
...
+        this.rulesList = Lists.newArrayList();
+
+        final List<Rule> presenceRules = Lists.newArrayList();
+        final List<Rule> exactRules = Lists.newArrayList();
+        final List<Rule> greaterRules = Lists.newArrayList();
+        final List<Rule> smallerRules = Lists.newArrayList();
+        final List<Rule> regexRules = Lists.newArrayList();
+
+        for (Stream stream : streams) {
+            final Boolean sufficient = stream.getMatchingType() == Stream.MatchingType.OR;
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: This could be a primitive `boolean`.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 3]
```diff
@@ -195,90 +231,24 @@ public String getFingerprint() {
...
-            }
-        }
-    }
-
-    private class Rule {
+    protected class Rule {
         private final Stream stream;
         private final StreamRule rule;
         private final StreamRuleMatcher matcher;
+        private final Boolean sufficient;
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: This could be a primitive `boolean`.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 4]
```diff
@@ -180,7 +216,7 @@ public String getFingerprint() {
 
             for (final StreamRule streamRule : stream.getStreamRules()) {
                 try {
-                    final Rule rule = new Rule(stream, streamRule);
+                    final Rule rule = new Rule(stream, streamRule, null);
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: What meaning would `sufficient = null` have in this case? There's also not a single check for `Rule#isSufficient() == null`.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 5]
```diff
@@ -83,33 +73,49 @@ public StreamRouterEngine(@Assisted List<Stream> streams,
         this.streamProcessingTimeout = streamFaultManager.getStreamProcessingTimeout();
         this.fingerprint = new StreamListFingerprint(streams).getFingerprint();
 
-        for (final Stream stream : streams) {
-            for (final StreamRule streamRule : stream.getStreamRules()) {
+        this.rulesList = Lists.newArrayList();
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: This list should be initialized with the (maximum) size `presenceRules.size()  + exactRules .size() + greaterRules.size() + smallerRules.size() + regexRules.size()`.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 6]
```diff
@@ -83,33 +73,49 @@ public StreamRouterEngine(@Assisted List<Stream> streams,
...
+
+        final List<Rule> presenceRules = Lists.newArrayList();
+        final List<Rule> exactRules = Lists.newArrayList();
+        final List<Rule> greaterRules = Lists.newArrayList();
+        final List<Rule> smallerRules = Lists.newArrayList();
+        final List<Rule> regexRules = Lists.newArrayList();
+
+        for (Stream stream : streams) {
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Maybe instead of sorting the stream rules manually, it would make sense to introduce some kind of complexity rating for the different types of rules, and use a comparator to sort the list of rules accordingly.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 7]
```diff
@@ -137,33 +143,63 @@ public String getFingerprint() {
      * @return the list of matching streams
      */
     public List<Stream> match(Message message) {
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Finally changing the method return type to `Set<Stream>` would make sense now, I guess. This would also avoid having to copy the matching result at the end of this method.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 8]
```diff
@@ -137,33 +143,63 @@ public String getFingerprint() {
...
+        Stream matchedStream = null;
+        try {
+            matchedStream = timeLimiter.callWithTimeout(new Callable<Stream>() {
+                @Override
+                public Stream call() throws Exception {
+                    return rule.match(message);
+                }
+            }, streamProcessingTimeout, TimeUnit.MILLISECONDS, true);
+        } catch (Exception e) {
+            streamFaultManager.registerFailure(rule.getStream());
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Logging the exception on DEBUG or TRACE level might help in case of errors.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 9]
```diff
@@ -195,90 +231,24 @@ public String getFingerprint() {
...
-            switch (stream.getMatchingType()) {
-                case AND: return ruleCount == matches;
-                case OR: return ruleCount > 0;
-                default: return ruleCount == matches;
-            }
-        }
-    }
-
-    private class Rule {
+    protected class Rule {
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Why did you change the visibility level to `protected`?

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 3 above)
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: While `sufficient` is shorter, I think `shortCircuit` or `shortCircuitRule` describes better, what this flag denotes in reality.

File: graylog2-server/src/test/java/org/graylog2/streams/StreamRouterEngineTest.java [hunk 10]
```diff
@@ -58,7 +59,7 @@ public void setUp() throws Exception {
         when(streamFaultManager.getStreamProcessingTimeout()).thenReturn(250L);
     }
 
-    private StreamRouterEngine newEngine(List<Stream> streams) {
+    protected StreamRouterEngine newEngine(List<Stream> streams) {
```
Comment (joschi) [inline graylog2-server/src/test/java/org/graylog2/streams/StreamRouterEngineTest.java]: Why was the visibility changed to `protected`?

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 11]
```diff
@@ -137,33 +143,63 @@ public String getFingerprint() {
...
+
+        for (Stream stream : result) {
+            streamMetrics.markIncomingMeter(stream.getId());
         }
 
-        return result;
+        return Lists.newArrayList(result);
+    }
+
+    private Stream matchWithTimeOut(final Message message, final Rule rule) {
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Please add `@Nullable`.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 12]
```diff
@@ -195,90 +231,24 @@ public String getFingerprint() {
...
             this.rule = rule;
+            this.sufficient = sufficient;
             this.matcher = StreamRuleMatcherFactory.build(rule.getType());
         }
 
+        public Boolean isSufficient() {
+            return sufficient;
+        }
+
         public Stream match(Message message) {
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Please add `@Nullable`.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 13]
```diff
@@ -137,33 +143,63 @@ public String getFingerprint() {
...
+            } else {
+                result.add(stream);
+                if (rule.isSufficient()) {
+                    // blacklist stream because it is already matched
+                    blackList.add(rule.getStream());
+                }
+            }
+        }
+
+
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Unnecessary blank line. :wink:

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 14]
```diff
@@ -137,33 +143,63 @@ public String getFingerprint() {
...
-        matchRules(message, Sets.intersection(fieldNames, smallerFields), smallerRules, matches);
-        // Execute regex rules with a timeout to prevent bad regexes to hang the processing.
-        matchRulesWithTimeout(message, Sets.intersection(fieldNames, regexFields), regexRules, matches, timeouts);
-
-        // Register failure for streams where rules ran into a timeout.
-        for (Stream stream : timeouts) {
-            streamFaultManager.registerFailure(stream);
-        }
+        final Set<Stream> result = Sets.newHashSet();
+        final Set<Stream> blackList = Sets.newHashSet();
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: These sets could be initialized with the (maximum) number of matches: [`Sets#newHashSetWithExpectedSize(int)`](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/Sets.html#newHashSetWithExpectedSize(int))

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 6 above)
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: I decided against this, because the effort to sort element by element is much higher than doing it manually.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 8 above)
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Done in the StreamFaultManager.registerFailure() method.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 14 above)
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Didn't do this, because it's misleading. We are not expecting it to have that size, it's is the maximum size they could have.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 4 above)
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: It is not relevant for the reuse of this class in the scope of the test matches, as shortcuts are not taken here and all rules need to be evaluated.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 7 above)
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Tried that. Not possible right now, as it would involve external changes of the API.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 4 above)
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: In this case, the `Rule#isSufficient()` method would need to be annotated with `@Nullable` and null-checks being introduced on every call of that method.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 8 above)
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: The exception is not being logged in `StreamFaultManager#registerFailure()`.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 14 above)
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: :white_check_mark:

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 7 above)
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: :white_check_mark:

Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: This is a bit unfortunate because right now this collection is at least copied twice for each message (one time in this method, another time in `Message#setStreams()`).

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 6 above)
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: :white_check_mark:

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 4 above)
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: TBH, I'd prefer simply using `false` here and making `Rule#isSufficient()` return a primitive `boolean`.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java [hunk 15]
```diff
@@ -83,33 +75,49 @@ public StreamRouterEngine(@Assisted List<Stream> streams,
...
+                        smallerRules.add(rule);
+                        break;
+                    case REGEX:
+                        regexRules.add(rule);
+                        break;
                 }
             }
         }
+
+        this.rulesList = new ImmutableList.Builder<Rule>()
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Unfortunately this suffers the same problem: Having to resize the underlying array multiple times if the number of rules is high (at least once if the number of rules is greater than 4).

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 7 above)
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Indeed. We should address this. It is beyond me why a List was chosen in the first place.

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 15 above)
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: I prefer immutability in this case. Remember that it's done only during initialization of the stream router (which is done only when the set of stream rules changes).

Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: :white_check_mark:

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 4 above)
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: > TBH, I'd prefer simply using `false` here and making `Rule#isSufficient()` return a primitive `boolean`.

Agreed, avoids null checks as well. Or is there a specific reason why you use `Boolean` instead of `boolean`?

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 8 above)
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Right!

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 4 above)
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: The reason was because of the reuse in the test matching logic. It is misleading to pass any value there. Null checking is a pita too. I will change to a primitive and add a comment.

Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Actually I will now replace it with the matching type of the stream. The wording was annoying from the beginning and a boolean flag is just not able to properly represent what it's supposed to do.

Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: :+1:

File: graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java (hunk 15 above)
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java]: Having looked at the surrounding code, this class is initialized every second (even if the rules haven't been changed) and `rulesList` is only used internally in this class, so I think initializing the list with the correct size would yield some improvements and is worth the disadvantage(?) over having a mutable collection instead of an immutable one.
